### PR TITLE
Convert search results to React

### PR DESCRIPTION
**Pull request by reosarevok**
This includes [MBS-9800](https://tickets.metabrainz.org/browse/MBS-9800), [MBS-9802](https://tickets.metabrainz.org/browse/MBS-9802), [MBS-9803](https://tickets.metabrainz.org/browse/MBS-9803), [MBS-9805](https://tickets.metabrainz.org/browse/MBS-9805), [MBS-9806](https://tickets.metabrainz.org/browse/MBS-9806), [MBS-9807](https://tickets.metabrainz.org/browse/MBS-9807), [MBS-9808](https://tickets.metabrainz.org/browse/MBS-9808), [MBS-9809](https://tickets.metabrainz.org/browse/MBS-9809), [MBS-9810](https://tickets.metabrainz.org/browse/MBS-9810), [MBS-9818](https://tickets.metabrainz.org/browse/MBS-9818), [MBS-9819](https://tickets.metabrainz.org/browse/MBS-9819), [MBS-9823](https://tickets.metabrainz.org/browse/MBS-9823)

### CODE REVIEW

File: root/types.js [hunk 1]
```diff
@@ -421,6 +421,7 @@ declare type PlaceT = {|
   ...CoreEntityRoleT,
   ...TypeRoleT<PlaceTypeT>,
   +entityType: 'place',
+  +address: string | null,
```
Comment (mwiencek) [inline root/types.js]: AFAICT this can't be `null` because it's marked as `NOT NULL` in the schema, and the Moose type in `Entity::Place` isn't `Maybe[str]`.

File: root/search/components/PlaceResults.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import loopParity from '../../utility/loopParity';
+import type {ResultsPropsT} from '../types';
+
+import PaginatedSearchResults from './PaginatedSearchResults';
+import ResultsLayout from './ResultsLayout';
+
+function buildResult(result, index) {
+  const place = result.entity;
+  const score = result.score;
+  console.log(place);
```
Comment (mwiencek) [inline root/search/components/PlaceResults.js]: Should be deleted

File: root/types.js [hunk 3]
```diff
@@ -421,6 +421,7 @@ declare type PlaceT = {|
   ...CoreEntityRoleT,
```
Comment (mwiencek) [inline root/types.js]: You also need to add `...DatePeriodRoleT,` after this line, and a line for `area:  AreaT | null,`.

To check the types locally and see if everything's fine, run `./node_modules/.bin/flow`

File: root/search/components/TagResults.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+import PaginatedSearchResults from './PaginatedSearchResults';
+import ResultsLayout from './ResultsLayout';
+
+function buildResult(result, index) {
+  const tag = result.entity;
+  const score = result.score;
+
+  return (
+    <tr className={loopParity(index)} data-score={score} key={tag}>
```
Comment (reosarevok) [inline root/search/components/TagResults.js]: Not sure what the purpose of key here is. It's usually ID, but the tags don't have an ID - they're strings. As such, I changed it to be the tag itself - let me know if I should just remove key completely instead.

Comment (yvanzo) [reply root/search/components/TagResults.js]: It is for React to identify an item in a list (of `tr`, `li`, `p`… actually any HTML tag).
See https://reactjs.org/docs/lists-and-keys.html#keys

File: root/search/components/ReleaseGroupResults.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        {l('Alternatively, you may {uri|add a new release group}.', {
+          __react: true,
+          uri: '/release_group/create?edit-release-group.name=' + encodeURIComponent(query),
+        })}
+      </p>
+    ) : null}
+  </ResultsLayout>
+);
+
+export default withCatalystContext(LabelResults);
```
Comment (mwiencek) [inline root/search/components/ReleaseGroupResults.js]: "ReleaseGroupResults"

File: root/search/components/ReleaseGroupResults.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      ]}
+      pager={pager}
+      query={query}
+      results={results}
+    />
+    {$c.user && !$c.user.is_editing_disabled ? (
+      <p>
+        {l('Alternatively, you may {uri|add a new release group}.', {
+          __react: true,
+          uri: '/release_group/create?edit-release-group.name=' + encodeURIComponent(query),
```
Comment (mwiencek) [inline root/search/components/ReleaseGroupResults.js]: "release-group" in the URI (replace underscore with hyphen)

File: root/search/components/InstrumentResults.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  const instrument = result.entity;
+  const score = result.score;
+
+  return (
+    <tr className={loopParity(index)} data-score={score} key={instrument.id}>
+      <td>
+        <EntityLink entity={instrument} />
+      </td>
+      <td>{instrument.typeName ? lp_attributes(instrument.typeName, 'instrument_type') : null}</td>
+      <td>{instrument.description}</td>
```
Comment (mwiencek) [inline root/search/components/InstrumentResults.js]: Some of the descriptions contain HTML, so you'll have to tell React by using `dangerouslySetInnerHTML` on the `<td>` (otherwise it'll display HTML to the user, since it escapes everything by default).

See how https://github.com/metabrainz/musicbrainz-server/blob/master/root/instrument/List.js#L35-L35 does it. Note it also uses `l_instrument_descriptions` for translation.

File: lib/MusicBrainz/Server/Entity/Recording.pm [hunk 8]
```diff
@@ -63,7 +63,7 @@ around TO_JSON => sub {
 
     return {
         %{ $self->$orig },
-        isrcs   => [map { $_->isrc } $self->all_isrcs],
+        isrcs   => [map { $_->TO_JSON } @{$self->isrcs}],
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Entity/Recording.pm]: Weird, I guess the Flow type was wrong all this time. But if you change this, you'll need to fix the usages of the `isrcs` property in root/static/scripts/common/MB/Control/Autocomplete.js and root/static/scripts/common/entity.js which expect a list of strings.

File: lib/MusicBrainz/Server/Entity/Area.pm [hunk 9]
```diff
@@ -103,10 +103,11 @@ around TO_JSON => sub {
...
 
     my $json = {
-        %{ $self->$orig },
-        iso_3166_1_codes => [$self->iso_3166_1_codes],
-        iso_3166_2_codes => [$self->iso_3166_2_codes],
-        iso_3166_3_codes => [$self->iso_3166_3_codes],
+        %{$self->$orig},
+        iso_3166_1_codes => [ $self->iso_3166_1_codes ],
+        iso_3166_2_codes => [ $self->iso_3166_2_codes ],
+        iso_3166_3_codes => [ $self->iso_3166_3_codes ],
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Entity/Area.pm]: You changed the code style here to add spaces inside the `[]` but remove spaces from inside the `%{}`, which doesn't seem consistent to me. :P (But if you want to change style, ideally do it in separate commits so that the functional changes are easier to see.)

File: root/types.js [hunk 10]
```diff
@@ -498,14 +499,34 @@ declare type ReleaseT = {|
...
   +entityType: 'release',
+  +events?: $ReadOnlyArray<ReleaseEventT>,
+  +labels?: $ReadOnlyArray<ReleaseLabelT>,
+  +language: LanguageT | null,
   +languageID: number | null,
   +packagingID: number | null,
+  +releaseGroup?: ReleaseGroupT,
+  +script: ScriptT | null,
   +scriptID: number | null,
+  +status: ReleaseStatusT,
```
Comment (mwiencek) [inline root/types.js]: Should be `ReleaseStatusT | null`.

File: root/search/components/WorkResults.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+import commaOnlyList from '../../static/scripts/common/i18n/commaOnlyList';
+import uniq from "lodash/uniq";
+
+
+function buildResult(result, index) {
+  const work = result.entity;
+  const score = result.score;
+
+  return (
+    <tr className={loopParity(index)} data-score={score} key={work.id}>
```
Comment (mwiencek) [inline root/search/components/WorkResults.js]: Cross-referencing https://github.com/metabrainz/musicbrainz-server/pull/722#discussion_r217522649 here.

File: root/types.js [hunk 12]
```diff
@@ -595,11 +600,23 @@ declare type WorkT = {|
...
+    +entity: ArtistT,
+    +roles: $ReadOnlyArray<string>,
+  |}>,
 |};
 
 export opaque type WorkTypeT: OptionTreeT = OptionTreeT;
 
+declare type WorkLanguageT = {|
+  +language: LanguageT,
+  +language_id: number,
```
Comment (mwiencek) [inline root/types.js]: `language_id` is no longer returned, so can be removed from here.

File: lib/MusicBrainz/Server/Entity/Area.pm (hunk 9 above)
Comment (reosarevok) [reply lib/MusicBrainz/Server/Entity/Area.pm]: Sigh. IDE again. Will undo. At this rate I might have to work on vim.

**Comment by mwiencek** [comment]
I force-pushed some changes, nothing major, but it was faster than leaving a bunch of comments since I was already reviewing each commit locally.

Some bugs were fixed, e.g. https://github.com/metabrainz/musicbrainz-server/pull/716/commits/121c5ab7912813db7884c1d65c307729798fab94 and with the CDStub search which wasn't working due to some broken exports. I also unmangled the last few commits (IIRC the doc and annotation changes were incorrectly squashed together into a commit for cdstubs) and made small style tweaks suggested by eslint.

It looks good to me now but there's a lot of code being converted here so @yvanzo should check too. :)
